gcloud-java-pubsub/src/main/java/com/google/cloud/pubsub/Message.java

```diff
@@ -0,0 +1,218 @@
+/*
+ * Copyright 2016 Google Inc. All Rights Reserved.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License"); you may not use this file except
+ * in compliance with the License. You may obtain a copy of the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software distributed under the License
+ * is distributed on an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express
+ * or implied. See the License for the specific language governing permissions and limitations under
+ * the License.
+ */
+
+package com.google.cloud.pubsub;
+
+import static com.google.common.base.Preconditions.checkNotNull;
+
+import com.google.common.base.MoreObjects;
+import com.google.common.collect.ImmutableMap;
+import com.google.protobuf.ByteString;
+import com.google.protobuf.Timestamp;
+import com.google.pubsub.v1.PubsubMessage;
+
+import java.io.Serializable;
+import java.util.HashMap;
+import java.util.Map;
+import java.util.Objects;
+
+/**
+ * Pub/Sub message.
+ */
+public class Message implements Serializable {
```

Comment:
To be clear, in the current proposal the Message keeps the content as ByteString. Therefore we don't do an extra copy in the translation layer.  So we are only talking about how to get/return the data from/to the user.

@mziccard Nice suggestions on how to *minimize* the cases where copy of content bytes is required.

My assumption is that PubSub messages tend not to be huge (I think they are restricted to less than 10MB) and typically consumed directly (as string or bytes) [please correct me if I am wrong]. `ByteString` already provides
various ways to set it (from String, InputStream,..) or to get its payload (String, byte[], Buffer,...). I think our options
in this case are:

1) Create our own `ByteString` like class that delegates to the protobuf `ByteString` (we did that in Datastore `Blob`). pros: not exposing proto, could be more flexible in input/output than proto. cons: still a copy when a user already have a `ByteString` (unfortunately it looks like `ByteString` still creates a copy if provided with a read-only ByteBuffer).

2) Provide various set payload methods in the Builder and get payload in the Message to support both efficient and convenient use-case. Suffer from the same con as (1)

3) Include `ByteString` as part of the api. pros: efficient, save work on duplicating in/out logic. cons: add new, external dependency in our API (less of an issue if that is something that we plan to do anyway  in the future),
less flexible in supporting more "convenient" method for getting/setting payload if needed.